### DISCUSSION TITLE
How to shutdown axum::serve with a CancellationToken?

### DISCUSSION THREAD

**Discussion by andinus** [tokio-rs/axum#2565] [Q&A]
### Summary

I have two tasks that I want to shutdown on `shutdown_signal()`. When I used `tokio::select` on the second task along with `shutdown_signal()`, I had to press `ctrl + c` twice, once to get axum to shutdown and then to shutdown the deletion task. 

I figured I could use a CancellationToken and cancel both tasks at once but with this Axum doesn't close connections gracefully. Am I missing something?

Here is the code:
![Screenshot 2024-02-04 at 3 32 39 PM](https://github.com/tokio-rs/axum/assets/62012039/d7685d23-d706-4ac3-ab35-e8025754aed3)

### axum version

0.7.4

**Comment by andinus** [comment]
`await` on both tasks at the end has fixed the issue. Here is the final code:
![Screenshot 2024-02-04 at 4 05 13 PM](https://github.com/tokio-rs/axum/assets/62012039/5ae6a085-3e7a-4f4c-9ae6-00940c574e22)

I'm not sure I understand why this works and why axum task is killed in original code. Is this the right way to do it?

**Reply by ndouglas** [reply]
Did you ever get more clarity on this? I'm having a doozy of a time getting Axum to shutdown cleanly.

**Reply by jplatte** [reply]
Have you seen the `graceful-shutdown` example in this repo? If yes but it didn't help, could you open a separate discussion explaining your specific problem in more detail?

**Reply by ndouglas** [reply]
Thank you! I'm just trying to figure out if I can create a minimal reproduction before I harass y'all 🙂
